### PR TITLE
RFR Er ss add reviews

### PR DESCRIPTION
**Pull request by cpsoinos**

### CODE REVIEW

File: app/controllers/reviews_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  def create
+    @restaurant = Restaurant.find(params[:restaurant_id])
+    @review = @restaurant.reviews.new(review_params)
+    @review.user = current_user
+    if @review.save
+      flash[:notice] = "Review Created!"
+      redirect_to restaurant_path(@restaurant)
+    else
+      render :new
```
Comment (omidbachari) [inline app/controllers/reviews_controller.rb]: Provide a flash if the save fails.

File: app/controllers/reviews_controller.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    if @review.user_id != current_user.id
+      flash[:notice] = "You cannot edit another user's review!"
+      render :"restaurants/show"
+    else
+      if @review.update(review_params)
+        flash[:notice] = "Review updated!"
+        redirect_to @restaurant
+      else
+        render :edit
```
Comment (omidbachari) [inline app/controllers/reviews_controller.rb:47]: Provide a flash if the update fails.

File: app/views/reviews/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<%= form_for [ @restaurant , @review ] do |f| %>
+<p>
+    <%= f.label :body %>
+    <%= f.text_field :body %>
+  </p>
+  <p>
+    <%= f.label :rating %> <br>
+    <% Review::RATINGS.each do |c| %>
+      <%= f.radio_button(:rating, c)%>
+      <%= f.label("rating_#{c.to_s}", c) %>
```
Comment (omidbachari) [inline app/views/reviews/_form.html.erb]: Was `.to_s` needed?

File: app/assets/stylesheets/reviews.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the reviews controller here.
```
Comment (dpickett) [inline app/assets/stylesheets/reviews.scss:1]: this file should be removed

File: app/assets/javascripts/reviews.coffee [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (dpickett) [inline app/assets/javascripts/reviews.coffee:1]: this file should be removed

File: app/controllers/reviews_controller.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
+class ReviewsController < ApplicationController
+  def new
+    @restaurant = Restaurant.find(params[:restaurant_id])
+    @review = Review.new
```
Comment (dpickett) [inline app/controllers/reviews_controller.rb:4]: a nitpick, but you could also do:

``` ruby
@review = @restaurant.reviews.build
```

File: app/controllers/reviews_controller.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    else
+      render :new
+    end
+  end
+
+  def destroy
+    @restaurant = Restaurant.find(params[:restaurant_id])
+    @review = Review.find(params[:id])
+
+    if @review.user_id == current_user.id
```
Comment (dpickett) [inline app/controllers/reviews_controller.rb:23]: consider writing an instance method for authorization.

``` ruby
@review.destroyable_by?(current_user)
```

File: app/controllers/reviews_controller.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  def edit
+    @restaurant = Restaurant.find(params[:restaurant_id])
+    @review = Review.find(params[:id])
+  end
+
+  def update
+    @review = Review.find(params[:id])
+    @restaurant = @review.restaurant
+
+    if @review.user_id != current_user.id
```
Comment (dpickett) [inline app/controllers/reviews_controller.rb:39]: consider an instance method for authorization here:

``` ruby
@review.editable_by?(current_user)
```

You should also authorize before the edit view is rendered

File: app/helpers/reviews_helper.rb [hunk 9]
```diff
@@ -0,0 +1,2 @@
+module ReviewsHelper
```
Comment (dpickett) [inline app/helpers/reviews_helper.rb:1]: this file can be removed

File: app/models/review.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class Review < ActiveRecord::Base
+  belongs_to :restaurant
+  belongs_to :user
+  validates :body, presence: true
+  validates :rating, presence: true
```
Comment (dpickett) [inline app/models/review.rb:5]: you should also validate that the rating is one of the defined elements in the `RATINGS` constant. Place the `RATINGS` constant above the validations, and use the `:inclusion` option

File: app/models/user.rb [hunk 11]
```diff
@@ -1,6 +1,9 @@
 class User < ActiveRecord::Base
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
+  has_many :reviews
```
Comment (dpickett) [inline app/models/user.rb:4]: what happens to reviews and restaurants if a user is destroyed?

File: app/views/restaurants/show.html.erb [hunk 12]
```diff
@@ -43,9 +43,25 @@
...
   <%= link_to 'Edit restaurant', edit_restaurant_path(@restaurant) %>
   <br />
   <%= link_to 'Delete restaurant', restaurant_path(@restaurant), method: :delete, data: { confirm: 'Are you sure?' } %>
 <% end %>
+
+<% if user_signed_in? %>
+  <%= link_to 'Write Review', new_restaurant_review_path(@restaurant) %>
+<% end %>
+
+<% @restaurant.reviews.each do |r| %>
```
Comment (dpickett) [inline app/views/restaurants/show.html.erb:58]: there is a scalability concern here. What happens if a restaurant has thousands of reviews? This list should be paginated.

File: app/views/restaurants/show.html.erb [hunk 13]
```diff
@@ -43,9 +43,25 @@
...
+  <p>Rating:<%=r.rating%></p>
+  <p>Review:<%=r.body%></p>
+  <% if r.user == current_user %>
```
Comment (dpickett) [inline app/views/restaurants/show.html.erb:61]: consider:

``` erb
<% if r.editable_by?(current_user) %>
```

File: app/views/restaurants/show.html.erb (hunk 12 above)
Comment (dpickett) [inline app/views/restaurants/show.html.erb:58]: consider an alternative local variable name

File: config/routes.rb [hunk 14]
```diff
@@ -1,9 +1,11 @@
...
-  [
+  root "restaurants#index"
+
+  resources :restaurants, only: [
     :index, :show, :new, :edit,
     :create, :update, :destroy
-  ]
-  root "restaurants#index"
+  ] do
+    resources :reviews, only: [:show, :create, :destroy, :edit, :new, :update]
```
Comment (dpickett) [inline config/routes.rb:9]: you can also say here `except: [:index]`

File: db/migrate/20150330221154_create_reviews.rb [hunk 15]
```diff
@@ -0,0 +1,12 @@
+class CreateReviews < ActiveRecord::Migration
+  def change
+    create_table :reviews do |t|
+      t.string :body
+      t.integer :rating
```
Comment (dpickett) [inline db/migrate/20150330221154_create_reviews.rb:5]: watch your schema constraints here. They should match what's specified in the model. you should have `null: false` here

File: db/migrate/20150330221154_create_reviews.rb [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+      t.references :restaurant, index: true, foreign_key: true
```
Comment (dpickett) [inline db/migrate/20150330221154_create_reviews.rb:6]: these should also be constrained. You can't have a review without a user or a restaurant

File: db/migrate/20150330221903_change_requirements_for_reviews_table.rb [hunk 17]
```diff
@@ -0,0 +1,19 @@
+class ChangeRequirementsForReviewsTable < ActiveRecord::Migration
+  def up
+    change_table :reviews do |t|
+      t.change :body, :string, {:null => false}
+      t.change :rating, :integer, {:null => false}
+      t.change :restaurant_id, :integer, {:null => false}
+      t.change :user_id, :integer, {:null => false}
```
Comment (dpickett) [inline db/migrate/20150330221903_change_requirements_for_reviews_table.rb:7]: good job - disregard last migrations commentary :-)

File: spec/models/review_spec.rb [hunk 18]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+# RSpec.describe Review, type: :model do
```
Comment (dpickett) [inline spec/models/review_spec.rb:3]: this file should be removed, or more ideally, you should have some unit tests for the review model

**Pull request merged**